### PR TITLE
Improve library targets

### LINKED ISSUE #6110: Throw an error for `"main": "something.html"` in package.json
# 🐛 bug report

It happens way too often that `"main": "index.html"` is specified in package.json, and the source files are overwritten.

## 🎛 Configuration (.babelrc, package.json, cli command)

```json
{
    "name": "foo",
    "version": "1.0.0",
    "main": "index.html",
    "license": "MIT",
    "private": true
}
```

## 🤔 Expected Behavior

`"main" musn't specified a HTML file. Remove this property`

## 😯 Current Behavior

Builds *something*

## 💁 Possible Solution

Should be checked in here: https://github.com/parcel-bundler/parcel/blob/v2/packages/core/core/src/requests/TargetRequest.js

### LINKED ISSUE #6465: Infer `"outputFormat": "esmodule"` for the "main" target with `"type": "module"`
With a `package.json` like
```json5
{
  "type": "module", // <----
  "main": "dist/index.js",
  "source": "src/index.js"
}
```

The default `outputFormat` of `main` should actually be `esmodule`, not `commonjs`.

### PR DESCRIPTION
**Pull request by devongovett**
Fixes #6110, Fixes #6465, Closes T-1091.

This adds a diagnostic when using an invalid file extension in one of the library targets.

image-2

In addition, it adds support for outputting `.mjs` and `.cjs` files, and infers the correct output format. The `"type": "module"` option in package.json is also supported.

### CODE REVIEW

File: packages/core/integration-tests/test/output-formats.js [hunk 1]
```diff
@@ -1108,6 +1108,40 @@ describe('output formats', function() {
       let ns = await run(b);
       assert.strictEqual(ns.fib(5), 8);
     });
+
+    it('should outputting .mjs files', async function() {
```
Comment (mischnic) [inline packages/core/integration-tests/test/output-formats.js]:
```suggestion
    it('should support outputting .mjs files', async function() {
```

**Comment by devongovett** [comment]
Should we also infer `outputFormat: 'esmodule'` or `outputFormat: 'commonjs'` for custom targets when using a `.mjs` or `.cjs` file? Currently I only infer this for common targets. I don't know if `type: module` should apply there as well? Does node only use it for the "main" field?

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 10.61s | -83.00ms |
| Cached | 453.00ms | -16.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/logo.1e014c76.png | 274.00b | +0.00b | 312.00ms | -37.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 7.53s | -30.00ms |
| Cached | 392.00ms | -21.00ms 🚀 |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/302609873616503304)**

**Comment by devongovett** [comment]
Added additional diagnostics:

* disallow global output format for library targets
* handle mismatch between declared output format and inferred
* disallow ESM output in `main` field without `.mjs` extension or `"type": "module"` field

**Review by joeyslater — APPROVED** [review]

**Pull request merged**